src/herder/TransactionQueue.cpp

```diff
@@ -216,20 +281,40 @@ TransactionQueue::shift()
     }
 }
 
-int
-TransactionQueue::countBanned(int index) const
+TransactionQueue::AddResult
+TransactionQueue::tryAdd(TransactionFrameBasePtr const& tx)
 {
-    return static_cast<int>(mBannedTransactions[index].size());
-}
+    if (isBanned(tx->getFullHash()))
+    {
+        return AddResult::ADD_STATUS_TRY_AGAIN_LATER;
+    }
 
-bool
-TransactionQueue::isBanned(Hash const& hash) const
-{
-    return std::any_of(
-        std::begin(mBannedTransactions), std::end(mBannedTransactions),
-        [&](std::unordered_set<Hash> const& transactions) {
-            return transactions.find(hash) != std::end(transactions);
-        });
+    auto pendingIter = mPendingTransactions.find(tx->getSourceID());
+    if (pendingIter != mPendingTransactions.end())
+    {
+        auto& transactions = pendingIter->second.mTransactions;
+        auto txIter = findTx(transactions, tx);
+        if (txIter != transactions.end())
```

Comment:
need comments here on what those cases are about (maybe renaming/better container choice for `mTransactions` will make this obvious).